File: app/src/main/java/com/iptvy/app/data/XtreamClient.kt
package com.iptvy.app.data

import android.util.JsonReader
import android.util.JsonToken
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

/**
 * Lightweight Xtream Codes client. No reflection-based JSON, minimal allocations,
 * per-category lazy loading so huge playlists never load all at once. Designed to
 * stay light on cheap Google TV sticks.
 */
class XtreamClient(private val prefs: Prefs) {

    private val http = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(20, TimeUnit.SECONDS)
        .retryOnConnectionFailure(true)
        .build()

    private fun apiUrl(action: String, extra: String = ""): String {
        val u = enc(prefs.username)
        val p = enc(prefs.password)
        return "${prefs.server}/player_api.php?username=$u&password=$p&action=$action$extra"
    }

    private fun enc(s: String): String = URLEncoder.encode(s, "UTF-8")

    private suspend fun getString(url: String): String = withContext(Dispatchers.IO) {
        val req = Request.Builder()
            .url(url)
            .header("User-Agent", "IPTVy/1.0")
            .build()
        http.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) throw RuntimeException("HTTP ${resp.code}")
            resp.body?.string() ?: ""
        }
    }

    /** Validates credentials. Returns true if the panel reports auth = 1. */
    suspend fun login(): Boolean {
        val body = getString("${prefs.server}/player_api.php?username=${enc(prefs.username)}&password=${enc(prefs.password)}")
        return try {
            val obj = JSONObject(body)
            val user = obj.optJSONObject("user_info")
            user != null && user.optInt("auth", 0) == 1
        } catch (e: Exception) {
            false
        }
    }

    suspend fun categories(type: StreamType): List<Category> {
        val action = when (type) {
            StreamType.LIVE -> "get_live_categories"
            StreamType.VOD -> "get_vod_categories"
            StreamType.SERIES -> "get_series_categories"
        }
        val arr = JSONArray(getString(apiUrl(action)))
        val out = ArrayList<Category>(arr.length() + 1)
        out.add(Category("__all__", "All"))
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            out.add(Category(o.optString("category_id"), o.optString("category_name")))
        }
        return out
    }

    /**
     * Streams the playlist with [JsonReader] straight off the socket — no full-body
     * String and no intermediate JSON tree, so even huge "All" catalogs stay light
     * on memory. The parsed [Stream] list is the only allocation that survives.
     */
    suspend fun streams(type: StreamType, categoryId: String): List<Stream> = withContext(Dispatchers.IO) {
        val action = when (type) {
            StreamType.LIVE -> "get_live_streams"
            StreamType.VOD -> "get_vod_streams"
            StreamType.SERIES -> "get_series"
        }
        val filter = if (categoryId == "__all__") "" else "&category_id=$categoryId"
        val req = Request.Builder()
            .url(apiUrl(action, filter))
            .header("User-Agent", "IPTVy/1.0")
            .build()
        http.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) throw RuntimeException("HTTP ${resp.code}")
            val body = resp.body ?: return@use emptyList<Stream>()
            JsonReader(body.charStream()).use { reader ->
                val out = ArrayList<Stream>()
                reader.beginArray()
                while (reader.hasNext()) out.add(readStream(reader, type))
                reader.endArray()
                out
            }
        }
    }

    private fun readStream(reader: JsonReader, type: StreamType): Stream {
        var streamId: String? = null
        var seriesId: String? = null
        var name = ""
        var icon: String? = null
        var cover: String? = null
        var ext: String? = null
        reader.beginObject()
        while (reader.hasNext()) {
            when (reader.nextName()) {
                "stream_id" -> streamId = reader.nextStringOrNull()
                "series_id" -> seriesId = reader.nextStringOrNull()
                "name" -> name = reader.nextStringOrNull().orEmpty()
                "stream_icon" -> icon = reader.nextStringOrNull()
                "cover" -> cover = reader.nextStringOrNull()
                "container_extension" -> ext = reader.nextStringOrNull()
                else -> reader.skipValue()
            }
        }
        reader.endObject()
        return when (type) {
            StreamType.LIVE -> Stream(streamId.orEmpty(), name, icon?.ifBlank { null }, StreamType.LIVE)
            StreamType.VOD -> Stream(
                streamId.orEmpty(), name, icon?.ifBlank { null }, StreamType.VOD,
                containerExt = ext?.ifBlank { null } ?: "mp4"
            )
            StreamType.SERIES -> Stream(seriesId.orEmpty(), name, cover?.ifBlank { null }, StreamType.SERIES)
        }
    }

    /** Reads the next value as a String (numbers stringified), or null for JSON null. */
    private fun JsonReader.nextStringOrNull(): String? =
        if (peek() == JsonToken.NULL) {
            nextNull()
            null
        } else {
            nextString()
        }

    /** Returns episodes grouped by season, ordered. */
    suspend fun seriesEpisodes(seriesId: String): List<Episode> {
        val body = getString(apiUrl("get_series_info", "&series_id=$seriesId"))
        val obj = JSONObject(body)
        val episodesObj = obj.optJSONObject("episodes") ?: return emptyList()
        val out = ArrayList<Episode>()
        val seasonKeys = episodesObj.keys().asSequence().toList().sortedBy { it.toIntOrNull() ?: 0 }
        for (key in seasonKeys) {
            val seasonNum = key.toIntOrNull() ?: 0
            val eps = episodesObj.optJSONArray(key) ?: continue
            for (i in 0 until eps.length()) {
                val e = eps.getJSONObject(i)
                out.add(
                    Episode(
                        id = e.opt("id").toString(),
                        title = e.optString("title").ifBlank { "Episode ${i + 1}" },
                        season = seasonNum,
                        containerExt = e.optString("container_extension").ifBlank { "mp4" }
                    )
                )
            }
        }
        return out
    }

    // ---- Stream URL builders ----

    fun liveUrl(streamId: String): String =
        "${prefs.server}/live/${enc(prefs.username)}/${enc(prefs.password)}/$streamId.ts"

    fun vodUrl(streamId: String, ext: String?): String =
        "${prefs.server}/movie/${enc(prefs.username)}/${enc(prefs.password)}/$streamId.${ext ?: "mp4"}"

    fun seriesUrl(episodeId: String, ext: String?): String =
        "${prefs.server}/series/${enc(prefs.username)}/${enc(prefs.password)}/$episodeId.${ext ?: "mp4"}"
}

File: desktop/lib/data/xtream_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

import 'models.dart';
import 'prefs.dart';

/// Xtream Codes client, ported from `XtreamClient.kt` / `xtream.js`. Same
/// `player_api.php` actions and the same /live, /movie, /series URL builders.
///
/// Unlike the browser build, a desktop app makes plain native HTTP requests, so
/// there is **no CORS and no proxy server** — every call goes straight to the
/// panel. That's the whole reason the web app needed a serverless `/api/proxy`
/// and this one doesn't.
class XtreamClient {
  final Prefs prefs;
  final http.Client _http;

  XtreamClient(this.prefs, {http.Client? httpClient})
      : _http = httpClient ?? http.Client();

  String _enc(String s) => Uri.encodeComponent(s);

  Uri _apiUrl(String action, [String extra = '']) => Uri.parse(
        '${prefs.server}/player_api.php?username=${_enc(prefs.username)}'
        '&password=${_enc(prefs.password)}&action=$action$extra',
      );

  Future<dynamic> _getJson(Uri url) async {
    final resp = await _http.get(url, headers: {
      'User-Agent': 'IPTVy/1.0',
      'Accept': 'application/json',
    });
    if (resp.statusCode < 200 || resp.statusCode >= 300) {
      throw Exception('HTTP ${resp.statusCode}');
    }
    return jsonDecode(utf8.decode(resp.bodyBytes));
  }

  /// Validates credentials. Returns true if the panel reports `auth == 1`.
  Future<bool> login() async {
    try {
      final url = Uri.parse(
        '${prefs.server}/player_api.php?username=${_enc(prefs.username)}'
        '&password=${_enc(prefs.password)}',
      );
      final obj = await _getJson(url);
      final userInfo = obj is Map ? obj['user_info'] : null;
      return userInfo is Map && _asInt(userInfo['auth']) == 1;
    } catch (_) {
      return false;
    }
  }

  Future<List<Category>> categories(StreamType type) async {
    final action = switch (type) {
      StreamType.live => 'get_live_categories',
      StreamType.vod => 'get_vod_categories',
      StreamType.series => 'get_series_categories',
    };
    final arr = await _getJson(_apiUrl(action));
    final out = <Category>[const Category('__all__', 'All')];
    if (arr is List) {
      for (final o in arr) {
        if (o is Map) {
          out.add(Category(
            (o['category_id'] ?? '').toString(),
            (o['category_name'] ?? '').toString(),
          ));
        }
      }
    }
    return out;
  }

  Future<List<Stream>> streams(StreamType type, String categoryId) async {
    final action = switch (type) {
      StreamType.live => 'get_live_streams',
      StreamType.vod => 'get_vod_streams',
      StreamType.series => 'get_series',
    };
    final filter =
        categoryId == '__all__' ? '' : '&category_id=${_enc(categoryId)}';
    final arr = await _getJson(_apiUrl(action, filter));
    if (arr is! List) return const [];
    return arr.whereType<Map>().map((o) => _mapStream(o, type)).toList();
  }

  Stream _mapStream(Map o, StreamType type) {
    switch (type) {
      case StreamType.series:
        return Stream(
          id: (o['series_id'] ?? '').toString(),
          name: (o['name'] ?? '').toString(),
          icon: _blankNull(o['cover']),
          type: type,
        );
      case StreamType.vod:
        return Stream(
          id: (o['stream_id'] ?? '').toString(),
          name: (o['name'] ?? '').toString(),
          icon: _blankNull(o['stream_icon']),
          type: type,
          containerExt: _blankNull(o['container_extension']) ?? 'mp4',
        );
      case StreamType.live:
        return Stream(
          id: (o['stream_id'] ?? '').toString(),
          name: (o['name'] ?? '').toString(),
          icon: _blankNull(o['stream_icon']),
          type: type,
        );
    }
  }

  /// Episodes flattened across seasons, in season then panel order.
  Future<List<Episode>> seriesEpisodes(String seriesId) async {
    final obj = await _getJson(_apiUrl('get_series_info', '&series_id=${_enc(seriesId)}'));
    final episodesObj = obj is Map ? obj['episodes'] : null;
    if (episodesObj is! Map) return const [];
    final seasonKeys = episodesObj.keys.map((e) => e.toString()).toList()
      ..sort((a, b) => (int.tryParse(a) ?? 0).compareTo(int.tryParse(b) ?? 0));
    final out = <Episode>[];
    for (final key in seasonKeys) {
      final seasonNum = int.tryParse(key) ?? 0;
      final eps = episodesObj[key];
      if (eps is! List) continue;
      for (var i = 0; i < eps.length; i++) {
        final e = eps[i];
        if (e is! Map) continue;
        final rawTitle = (e['title'] ?? '').toString().trim();
        final rawExt = (e['container_extension'] ?? '').toString().trim();
        out.add(Episode(
          id: (e['id'] ?? '').toString(),
          title: rawTitle.isNotEmpty ? rawTitle : 'Episode ${i + 1}',
          season: seasonNum,
          containerExt: rawExt.isNotEmpty ? rawExt : 'mp4',
        ));
      }
    }
    return out;
  }

  // ---- Stream URL builders ----

  String liveUrl(String streamId) =>
      '${prefs.server}/live/${_enc(prefs.username)}/${_enc(prefs.password)}/$streamId.ts';

  String vodUrl(String streamId, String? ext) =>
      '${prefs.server}/movie/${_enc(prefs.username)}/${_enc(prefs.password)}/$streamId.${ext ?? 'mp4'}';

  String seriesUrl(String episodeId, String? ext) =>
      '${prefs.server}/series/${_enc(prefs.username)}/${_enc(prefs.password)}/$episodeId.${ext ?? 'mp4'}';

  void dispose() => _http.close();

  static int _asInt(dynamic v) =>
      v is int ? v : (v is String ? int.tryParse(v) ?? 0 : 0);

  static String? _blankNull(dynamic v) {
    if (v == null) return null;
    final s = v.toString();
    return s.trim().isEmpty ? null : s;
  }
}
